proxy/proxy.go

```diff
@@ -72,10 +75,64 @@ type Proxy struct {
 	weaveWaitNoopVolume string
 	normalisedAddrs     []string
 	waiters             map[*http.Request]*wait
+	attachJobs          map[string]*attachJob
+	quit                chan struct{}
+}
+
+type attachJob struct {
+	id          string
+	tryInterval time.Duration // retry delay on next failure
+	timer       *time.Timer
+}
+
+func (proxy *Proxy) attachWithRetry(id string) {
+	proxy.Lock()
+	defer proxy.Unlock()
+	if j, ok := proxy.attachJobs[id]; ok {
+		j.timer.Reset(time.Duration(0))
+		return
+	}
+
+	j := &attachJob{id: id, tryInterval: initialInterval}
+	j.timer = time.AfterFunc(time.Duration(0), func() {
+		if err := proxy.attach(id, false); err != nil {
+			j.reschedule()
+			return
+		}
+		proxy.notifyWaiters(id, nil)
+	})
+	proxy.attachJobs[id] = j
```

Comment:
I found it quite hard to figure out where the "retry loop" is.  The payload of `time.AfterFunc` gets called, calls `reschedule`, which calls `timer.Reset`, which causes the payload to be called again later.

It might be slightly easier to follow if the `Reset` call was inside the payload.